joomla/joomla-cms
# Improvements for secure registration and login

This is a revised version of PR #8619 based upon the current Joomla staging version.

Assume the following setup:
- the Joomla server is reachable via HTTP and HTTPS
- the Joomla administration is accessed via HTTPS ("Force HTTPS" is set to "Administrator Only")
- the Joomla site shall be accessed by default via HTTP
- optionally the ports used for HTTP and HTTPS are non-standard and shall be configurable

This PR aims at the following:
- there should be a single place for activating secure login and registration
- user registration data shall be transferred via HTTPS to the server if activated
- user login at the frontend shall always use HTTPS if activated
- all pages shown when logged in at the frontend shall always be transferred via HTTPS if activated
- user logout at the frontend shall always return to HTTP if secure login is activated
- site preview from the administration shall always open the site home page in HTTP mode
- switching between HTTP and HTTPS shall work even when non-standard ports are used
- if "Force HTTPS" is set to None no switching between HTTP and HTTPS shall be done

Currently secure login can be activated in each `mod_login` instance separately (and there are at least three of them in the demo site delivered with Joomla), and there is no way to enable secure login when using the login form menu item coming from `com_users`. Moreover it is not possible to activate secure transfer for user registration data which include sensitive data like password and mail address. Also the reminding and resetting functions shall be able to use HTTPS for transferring e.g. the mail address (currently HTTP is used for that).

In order to achieve this I moved the "usesecure" setting from the `mod_login` instance settings to the `com_users` configuration, and I changed all affected files (including message files) accordingly. In order to make foreign templates that override the `mod_login` login form work as before (assuming they check for the "usesecure" setting when logging in) I copy the "usesecure" setting from the `com_users` configuration dynamically to the `mod_login` instance settings prior to activating the login form of the template.

For proper links from the backend running in HTTPS mode to the site (site preview) I modified the administrator templates hathor and isis accordingly, moreover I had to change the administrator module `mod_status` for that purpose.

In order to handle internal URL checking even in the case of switching between HTTP and HTTPS I had to change the `JUri::isInternal()` method.

I order to make switching between HTTP and HTTPS protocols easier (same lines of code needed at many places) I implemented two new methods `JUri::siteScheme()` and `JUri::adminScheme()` which do this switching.

If you enable "Force HTTPS" by setting it to either "Administrator Only" or "Entire Site" (Global Configuration, Server Settings) you'll now get two additional fields for the HTTP and the HTTPS ports where you can enter the ports used at your site. This is necessary only when using non-standard ports like 8080 for HTTP and 8443 for HTTPS.

If you change these ports they are checked as well as the "Force HTTPS" mode itself, and in case of an error doing so the change is undone.

At all code locations where switching between HTTP and HTTPS is performed these ports if set are used accordingly.

Note that switching between HTTP and HTTPS protocols is done only if "Force HTTPS" is set to anything else than "None". This should help in cases like in PR #11213.

The login and logout forms in `mod_login` used the current URL as their target for the forms which is of no great use since the real target of these forms is passed in the hidden field "return" I changed that back to "index.php" as it was about two years ago.

I saw that the beez3 template contains an override for the `mod_login` login form but no separate override for the logout form. Instead the login form still contained the login and logout cases as was the case in `mod_login` until more than three years ago (August 2012).

Finally I added the new field attribute "renderon" (see PR #11207) both to the "usesecure" field in `com_users` and to the "secure" field in `com_menus`, these fields are shown only if the global flag "force_ssl" is set to 1 ("Administrator Only") and if PR #11207 has been applied. My reason for this is that if "force_ssl" is 0 probably no HTTPS is available such that switching to HTTPS would most likely fail, and if "force_ssl" is 2 ("Entire Site") all pages are already shown via HTTPS such that no switching is necessary.
#### Testing Instructions
-  set up current Joomla version including demo site and apply all changes from this PR
- optionally apply the changes from PR #11207 
- the webserver must be reachable via HTTP and HTTPS
- log in to the backend as administrator
- set "Force HTTPS" in Global Configuration, Server Settings to "Administrator Only"
- the backend should now be accessed via HTTPS only
- all site preview links should point to HTTP (hathor or isis template assumed)
- frontend login (regardless which) continue in HTTP mode
- go to the Users Configuration and enable "Encrypt Login / Registration Forms"
- check registration form: the action URL should point to HTTPS
- after successful registration the user will be back in HTTP mode
- check frontend login form (regardless which): the action URL should point to HTTPS
- after successful login all pages should be displayed via HTTPS
- in case of a login failure the user should be back in HTTP mode
- check frontend logout form: the action URL should point to HTTPS
- perform logout: the user should be back in HTTP mode
- check remind and reset links: the action URL for both of them should point to HTTPS
- after performing remind / reset functions the user should be back in HTTP mode
- set any menu item to "Secure" mode (tab "Metadata")
- check in the frontend that this menu item now points to HTTPS
- configure additional server ports usable for HTTP and HTTPS, e.g. to 8080 and 8443
- configure these ports in Global Configuration, Server Settings (below "Force HTTPS")
- check that switching between HTTP and HTTPS works as expected
- check that the action URLs mentioned above contain the right HTTPS port
- set "Force HTTPS" in Global Configuration, Server Settings to "Entire Site"
- all pages should now be served via HTTPS, no switching when logging in or out
- if PR #11207 was applied too:
  - the field "Encrypt Login / Registration Forms" in the Users Configuration is no longer rendered
  - the field "Secure" in menu items (tab "Metadata") is no longer rendered

Note: In testing this PR you might encounter that your browser uses the HTTPS protocol even if you changed "Force HTTPS" back from "Entire Site" to "Administrator Only" or "None" or back from "Administrator Only" to "None". This is not a problem with these changes but is caused by the browser caching redirections it received prior to changing the "Force HTTPS" setting. In such cases please close all windows or tabs showing content from the Joomla instance used for testing, clear the browser's cache and then try again.
